In [5]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import KFold, cross_val_score
from sklearn.linear_model import Ridge
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

data = pd.read_csv("A2data.tsv", sep="\t")
data = data.iloc[:, 1:]
X = data.iloc[:, :-1]
y = data.iloc[:, -1]
model = LinearRegression()
cv = KFold(n_splits=5, shuffle=True, random_state=42)
scores = cross_val_score(model,X,y,cv=cv,scoring='neg_root_mean_squared_error')
root_mean_squared_error = -scores
print("Baseline Model: Linear Regression")
print("RMSE for each fold:", root_mean_squared_error)
print("Mean RMSE:", root_mean_squared_error.mean())
print("Std RMSE:", root_mean_squared_error.std())

Baseline Model: Linear Regression
RMSE for each fold: [1.47526526 1.888405   1.8315792  1.94119887 2.81917886]
Mean RMSE: 1.9911254390867021
Std RMSE: 0.44503558269735577


Task 1 builds a baseline regression model using Linear Regression on the raw dataset. The first column (InstanceID) was removed because it contains string identifiers and cannot be used as numeric input. The remaining 99 features were used to predict the output variable Y.

A 5-fold cross-validation strategy was chosen because the dataset contains only 48 instances, which is relatively small. Using 5 folds allows the model to be trained on most of the data while still providing multiple independent test evaluations, giving a good balance between reliable performance estimation and sufficient training data in each fold.

Model performance was evaluated using RMSE (Root Mean Squared Error), which measures the average magnitude of prediction error. The baseline model achieved an average RMSE of 1.99 ± 0.45, indicating that Linear Regression provides a reasonable initial benchmark for comparison with alternative regression methods.


In [6]:
model = Pipeline([('scaler', StandardScaler()),('ridge', Ridge(alpha=1.0))])
cv = KFold(n_splits=5, shuffle=True, random_state=42)
scores = cross_val_score(model,X,y,cv=cv,scoring='neg_root_mean_squared_error')
root_mean_squared_error = -scores
print("Alternative Model 1: Ridge Regression")
print("RMSE for each fold:", root_mean_squared_error)
print("Mean RMSE:", round(root_mean_squared_error.mean(), 4))
print("Std RMSE:", round(root_mean_squared_error.std(), 4))

Alternative Model 1: Ridge Regression
RMSE for each fold: [1.27497298 1.36958043 1.27309185 1.87142235 1.39072096]
Mean RMSE: 1.436
Std RMSE: 0.2229


As an alternative model, Ridge Regression was selected because the dataset contains many predictors relative to the number of observations. Ridge adds L2 regularization to the regression objective, which helps reduce overfitting by shrinking coefficient values.

Standardization was applied before training because Ridge Regression is sensitive to feature scale. A 5-fold cross-validation strategy was used to evaluate performance consistently with the baseline model.
